doc/developer/design/20231113_optimizer_notice_catalog.md

```diff
@@ -0,0 +1,480 @@
+   some idea how these will work out).
+
+## Solution Proposal
+
+<!--
+What is your preferred solution, and why have you chosen it over the
+alternatives? Start this section with a brief, high-level summary.
+
+This is your opportunity to clearly communicate your chosen design. For any
+design document, the appropriate level of technical details depends both on
+the target reviewers and the nature of the design that is being proposed.
+A good rule of thumb is that you should strive for the minimum level of
+detail that fully communicates the proposal to your reviewers. If you're
+unsure, reach out to your manager for help.
+
+Remember to document any dependencies that may need to break or change as a
+result of this work.
+-->
+
+3. [Console UX and Workflows](#console-ux-and-workflows) sketches the main UX
+   workflows that the current design aims to support.
+1. [Catalog Schema](#catalog-schema) section describes the design of a catalog
+   schema that contains the optimizer notices corresponding to the current
+   catalog state.
+2. [Schema Maintenance](#schema-maintenance) elaborates how the contents of the
+   proposed catalog schema will be maintained at runtime, including the
+   mechanisms mechanisms used to align the (externally visible) status of a
+   notice with the internal system state.
+
+### Console UX and Workflows
+
+This section contains wireframe diagrams that illustrate the console UX and
+workflows that will be supported by the proposed notices schema. Each wireframe
+links to [an Excalidraw scene][excalidraw_sources] which contains the wireframe
+sources and supports leaving comments directly on the diagram.
+
+#### Listing all Notices
+
+![Listing all objects wireframe](./static/20231113_optimizer_notice_catalog/ux_listing_all_notices.png)
+
+- For the MVP, all notices will be listed in a single page. In the future, we
+  can have scoped versions of this view as a tab on the clusters page.
+- The customer will be able to change the status of a notice directly from the
+  list view:
+  - A notice marked as "resolved" indicates that the customer has performed the
+    suggested corrective correction. This does not imply that the notice is
+    already resolved, but that we expect this to happen during the next once the
+    related catalog objects have been (re-)created.
+  - A notice marked as "ignored" indicates that the user considers it irrelevant
+    or low priority and they don't want to see it the foreseeable future.
+- The page will contain an extra filter dropdown that will allow customers to
+  filter out notices with a certain status:
+  - hide ignored and resolved - probably this should be the default,
+  - hide ignored - probably most useful for field engineers using the
+    impersonated console,
+  - show all.
+
+#### Inspecting a Notice
+
+![Listing all objects wireframe](./static/20231113_optimizer_notice_catalog/ux_inspecting_a_notice.png)
```

Comment:
(For the corrective action of IndexTooWideForLiteralConstraints, we should mention that the user might want to drop the old index _if it is not used by some other queries_.)

Reply:
Are you with OK with just emitting a notice for indexes that are not used at all? Then handling the `IndexTooWideForLiteralConstraints` and re-creating the objects might result in another notice.

Reply:
This is tricky, because we'd like to keep notices quite high-signal, i.e., false positives should be rare. But `IndexNotUsed` would definitely show a false positive for each serving index, if we implement it by only looking at static objects. So, I'd say we should add this notice only when we'll also have the last usage time of each index implemented (https://github.com/MaterializeInc/materialize/issues/20877), because then we can take that also into account when deciding whether to show an `IndexNotUsed`.

Reply:
> Are you with OK with just emitting a dedicated notice for indexes that are not used at all?

Ok! And then when we have `IndexNotUsed` later implemented in the above way, then these things will fall into place.

Reply:
> So, I'd say we should add this notice only when we'll also have the last usage time of each index implemented (https://github.com/MaterializeInc/materialize/issues/20877), because then we can take that also into account when deciding whether to show an IndexNotUsed.

The design also provisions for ignoring a notice in a way that will survive restarts. If this is the case the user can just manually ignore all false positive notices for `IndexNotUsed`.